shrubbery-lib/shrubbery/lex.rkt

```diff
@@ -1055,29 +1055,34 @@
    [(:+ whitespace) (values lexeme end-pos)]))
 
 (define (parse-number s)
-  (if (and ((string-length s) . > . 2)
-           (eqv? #\0 (string-ref s 0)))
-      (case (string-ref s 1)
-        [(#\x) (string->number (regexp-replace* #rx"_" (substring s 2) "") 16)]
-        [(#\o) (string->number (regexp-replace* #rx"_" (substring s 2) "") 8)]
-        [(#\b) (string->number (regexp-replace* #rx"_" (substring s 2) "") 2)]
-        [else
-          (string->number (regexp-replace* #rx"_" s ""))])
-      (string->number (regexp-replace* #rx"_" s ""))))
+  (define-values (negative? radix num)
+    (cond
+      [(regexp-match #rx"^([+-])?(0[xob])?(.+)$" s)
```

Comment:
I worry a little about the cost of the redundant parsing here, since an extra layer of regexp matching and allocating string matches is a lot more expensive than the old character tests. To check whether it could possibly matter, I tried the program below before an after pulling the commit. It reports 25ms before and 29ms after on my machine, which seems significant given everything else that happens for parsing.

```
#lang rhombus
import:
  rhombus/shrubbery
  rhombus/measure

def s = String.append(& for List (i: 0..10000):
                        " " +& i)
def p = Port.Input.open_string(s)

measure.time:
  ~gc
  shrubbery.read(p)
  #void
```

Reply:
Updated to address the concern, hopefully.